File: rxbinding-design/src/main/java/com/jakewharton/rxbinding2/support/design/widget/AppBarLayoutOffsetChangeObservable.java
package com.jakewharton.rxbinding2.support.design.widget;

import android.support.design.widget.AppBarLayout;
import android.support.design.widget.AppBarLayout.OnOffsetChangedListener;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class AppBarLayoutOffsetChangeObservable extends Observable<Integer> {
  private final AppBarLayout view;

  AppBarLayoutOffsetChangeObservable(AppBarLayout view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super Integer> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.addOnOffsetChangedListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnOffsetChangedListener {
    private final AppBarLayout appBarLayout;
    private final Observer<? super Integer> observer;

    Listener(AppBarLayout appBarLayout, Observer<? super Integer> observer) {
      this.appBarLayout = appBarLayout;
      this.observer = observer;
    }

    @Override public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
      if (!isDisposed()) {
        observer.onNext(verticalOffset);
      }
    }

    @Override protected void onDispose() {
      appBarLayout.removeOnOffsetChangedListener(this);
    }
  }
}

File: rxbinding-material/src/main/java/com/jakewharton/rxbinding3/material/AppBarLayoutOffsetChangeObservable.kt
@file:JvmName("RxAppBarLayout")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.material

import androidx.annotation.CheckResult
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.AppBarLayout.OnOffsetChangedListener
import com.jakewharton.rxbinding2.internal.checkMainThread
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

/**
 * Create an observable which emits the offset change in `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun AppBarLayout.offsetChanges(): Observable<Int> {
  return AppBarLayoutOffsetChangeObservable(this)
}

private class AppBarLayoutOffsetChangeObservable(
  private val view: AppBarLayout
) : Observable<Int>() {

  override fun subscribeActual(observer: Observer<in Int>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.addOnOffsetChangedListener(listener)
  }

  private class Listener(
    private val appBarLayout: AppBarLayout,
    private val observer: Observer<in Int>
  ) : MainThreadDisposable(), OnOffsetChangedListener {

    override fun onOffsetChanged(appBarLayout: AppBarLayout, verticalOffset: Int) {
      if (!isDisposed) {
        observer.onNext(verticalOffset)
      }
    }

    override fun onDispose() {
      appBarLayout.removeOnOffsetChangedListener(this)
    }
  }
}
